// Copyright 2025 The llm-d Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package tokenization;

option go_package = "github.com/llm-d/llm-d-kv-cache/api/tokenizerpb;tokenizerpb";

// ChatMessage represents a single message in a conversation
message ChatMessage {
  string role = 1;    // Role of the message (e.g., "user", "assistant", "system")
  string content = 2; // Content of the message
}


// Value represents a generic value that can be string, number, bool, or list
message Value {
  oneof value {
    string string_value = 1;
    double number_value = 2;
    bool bool_value = 3;
    ListValue list_value = 4;
    StructValue struct_value = 5;
  }
}

// ListValue represents a list of values
message ListValue {
  repeated Value values = 1;
}

// StructValue represents a structured value (key-value pairs)
message StructValue {
  map<string, Value> fields = 1;
}

// InitializeTokenizerRequest represents a request to initialize tokenizer for a model
message InitializeTokenizerRequest {
  bool is_local = 1;          // Whether the model is local (default: true)
  string model = 2;           // The model ID or path (HF model ID, local directory path, or path to tokenizer file)
  optional string revision = 3;       // Model revision (optional)
  optional string token = 4;          // Hugging Face token for private models (optional)
  optional string download_dir = 5;   // Directory to download the model (optional)
}

// InitializeTokenizerResponse represents the response from tokenizer initialization
message InitializeTokenizerResponse {
  bool success = 1;           // Whether the initialization was successful
  string error_message = 2;   // Error message if initialization failed
}

// RenderRequest represents a request to render (tokenize) a text input
message RenderRequest {
  string text = 1;              // The text input to render/tokenize
  string model_name = 2;        // The name of the model to use for tokenization
  bool add_special_tokens = 3;  // Whether to add special tokens during tokenization
}

// RenderChatRequest represents a request to render a chat conversation
message RenderChatRequest {
  repeated ChatMessage conversation = 1;  // The conversation messages (array of role/content pairs)
  repeated Value tools = 2;               // Tools available to the conversation (arbitrary values)
  repeated Value documents = 3;           // Documents related to the conversation (arbitrary values)
  optional string chat_template = 4;               // The chat template to use
  optional bool return_assistant_tokens_mask = 5;  // Whether to return assistant token mask
  optional bool continue_final_message = 6;        // Whether to continue the final message
  optional bool add_generation_prompt = 7;         // Whether to add generation prompt
  map<string, Value> chat_template_kwargs = 8; // Additional chat template arguments
  string model_name = 9;                  // The name of the model to use for tokenization
}

// RenderResponse represents the response from rendering/tokenizing
message RenderResponse {
  repeated uint32 input_ids = 1;           // Token IDs for the input
  bool success = 2;                        // Whether the request was successful
  string error_message = 3;                // Error message if the request failed
  // Direct array of [start, end] pairs
  repeated uint32 offset_pairs = 4;        // Flattened array of [start, end, start, end, ...]
}

// TokenizationService defines the gRPC service for tokenization
service TokenizationService {
  // InitializeTokenizer initializes the tokenizer for a specific model
  rpc InitializeTokenizer(InitializeTokenizerRequest) returns (InitializeTokenizerResponse);

  // Render renders (tokenizes) a text input
  rpc Render(RenderRequest) returns (RenderResponse);

  // RenderChat renders a chat conversation to tokens and offsets
  rpc RenderChat(RenderChatRequest) returns (RenderResponse);
}